libraries/libstratosphere/source/settings/impl/settings_key_value_store.cpp

```diff
@@ -0,0 +1,1629 @@
+
+    Result ReadKeyValueStorePlatformConfiguration(u64 *out_count, char * const out_buffer, size_t out_buffer_size) {
+        /* Check preconditions. */
+        AMS_ASSERT(out_count != nullptr);
+        AMS_ASSERT(out_buffer != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Read the platform configuration system data. */
+        return ReadSystemDataPlatformConfiguration(out_count, out_buffer, out_buffer_size);
+    }
+
+    Result ReadKeyValueStoreSaveData(u64 *out_count, char * const out_buffer, size_t out_buffer_size) {
+        /* Check preconditions. */
+        AMS_ASSERT(out_count != nullptr);
+        AMS_ASSERT(out_buffer != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Read the system save data. */
+        return ReadSystemSaveData(out_count, out_buffer, out_buffer_size);
+    }
+
+    Result ReloadKeyValueStoreForDebug(SystemSaveData *system_save_data, SystemSaveData *fwdbg_system_data, SystemSaveData *pfcfg_system_data) {
+        /* Check preconditions. */
+        AMS_ASSERT(system_save_data != nullptr);
+        AMS_ASSERT(fwdbg_system_data != nullptr);
+        AMS_ASSERT(pfcfg_system_data != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMapForciblyForDebug(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Load the key value store map. */
+        return LoadKeyValueStoreMapForDebug(map, system_save_data, fwdbg_system_data, pfcfg_system_data);
+    }
+
+    Result ReloadKeyValueStoreForDebug() {
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMap(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Load the key value store map. */
+        return LoadKeyValueStoreMap(map);
+    }
+
+    Result ResetKeyValueStoreSaveData() {
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMap(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Reset all values in the map. */
+        for (auto it = map->begin(); it != map->end(); ++it) {
+            /* Get the map value. */
+            MapValue &map_value = it->second;
+
+            /* If the current value isn't the default value, reset it. */
+            if (map_value.current_value != map_value.default_value) {
+                /* Free the current value if present. */
+                if (map_value.current_value != nullptr) {
```

Comment:
Nintendo updates the map value *before* freeing....do this using temporaries to get current, then setting current = defualt, then freeing temporaries if not null?